### PR TITLE
Added filters and transforms to table definitions

### PR DESCRIPTION
**Pull request by danderson00**
For example:

```Javascript
mobileApp.tables.add('tableName', {
    filters: [function (query, context) {
        query.where({ someProperty: context.req.get('some-header') })
    }],
    transforms: [function (item, context) {
        item.someProperty = context.req.get('some-header')
    }]
})
```

Per-user data implementation based on this, turned on with a flag to come in another PR.

Interestingly, there are strong parallels between what we are doing here and middleware, and this would not be required at all if we used middleware for things like per user data. The problem with middleware is that it doesn't get executed when we say `context.tables('otherTable').read()` from table operations or custom APIs.

@paulbatum We should talk about the above point.

Also @mamaso @fabiocav @shrishrirang for code / API review

### CODE REVIEW

File: src/data/mssql/index.js [hunk 1]
```diff
@@ -19,8 +19,6 @@ module.exports = function (configuration) {
     setEncryption();
 
     var tableAccess = function (table) {
```
Comment (danderson00) [inline src/data/mssql/index.js]: Just removing all the argument assertions here. They are now all moved to wrapProvider.

File: src/data/wrapProvider.js [hunk 2]
```diff
@@ -0,0 +1,58 @@
+// ----------------------------------------------------------------------------
```
Comment (danderson00) [inline src/data/wrapProvider.js]: wrapProvider handles asserting all the required arguments are present and applying filters and transforms.

File: src/express/middleware/createContext.js [hunk 3]
```diff
@@ -30,7 +30,7 @@ module.exports = function (configuration) {
             tables: function (name) {
                 if(!configuration.tables[name])
                     throw new Error("The table '" + name + "' does not exist.")
-                return attachOperators(name, dataProvider(configuration.tables[name]));
+                return attachOperators(name, dataProvider(configuration.tables[name], context));
```
Comment (danderson00) [inline src/express/middleware/createContext.js]: The data provider table access function now accepts a context parameter. Pass it in here and below in the table operation mappings.

File: test/express/integration/tables/filters.tests.js [hunk 4]
```diff
@@ -1,114 +1,51 @@
 // ----------------------------------------------------------------------------
 // Copyright (c) Microsoft Corporation. All rights reserved.
 // ----------------------------------------------------------------------------
-﻿var expect = require('chai').use(require('chai-subset')).expect,
+var expect = require('chai').use(require('chai-subset')).expect,
```
Comment (danderson00) [inline test/express/integration/tables/filters.tests.js]: Just renaming filters.tests.js to a more appropriate operations.tests.js

File: src/express/tables/operations.js [hunk 5]
```diff
@@ -6,21 +6,21 @@
 module.exports = {
```
Comment (danderson00) [inline src/express/tables/operations.js]: haha yeh, I cringed a little when I added the context parameter. I'll leave this file tucked away for some future maintainer to enjoy :D

File: test/express/integration/tables/filters.tests.js (hunk 4 above)
Comment (mamaso) [inline test/express/integration/tables/filters.tests.js]: I wonder if you can give git the hint that the original file is renamed

**Comment by mamaso** [comment]
looks good.  filters & transforms seems like a good way to add pluggable components

**Comment by shrishrirang** [comment]
Don't have enough context. I am skipping this PR.

**Comment by danderson00** [comment]
@shrishrirang A general code quality / API structure pass would be great.

File: src/data/index.js [hunk 6]
```diff
@@ -18,9 +20,25 @@ object with the members described below. The function also has an
 be used to execute raw SQL queries.
 */
 module.exports = function (configuration) {
-    var provider = (configuration && configuration.data && configuration.data.provider) || 'memory';
-    return (types.isFunction(provider) ? provider : require('./' + provider))(configuration.data);
-}
+    var provider = createProvider();
+
+    var api = function (table, context) {
+        assert(table, 'A table was not specified');
```
Comment (shrishrirang) [inline src/data/index.js]: should we do little more validation? checking that the table is a non-empty string or whatever type we expect here?

File: src/data/wrapProvider.js [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// ----------------------------------------------------------------------------
+var assert = require('../utilities/assert').argument,
+    queries = require('../query');
+
+module.exports = function (provider, table, context) {
```
Comment (shrishrirang) [inline src/data/wrapProvider.js]: 'wrapProvider' doesn't quite sound a great name to me, but I don't have a great suggestion. Will 'providerFilter' or something alike convey what it does better? Optional comment.

File: src/data/index.js (hunk 6 above)
Comment (danderson00) [inline src/data/index.js]: This is generally not consumed by user code. The parameter type is actually a table definition object, which can potentially just be an empty object, so more validation is not particularly valuable.

File: src/data/wrapProvider.js (hunk 7 above)
Comment (danderson00) [inline src/data/wrapProvider.js]: Yeh, this was the best I could come up with. I previously called it filters.js, but it's more than that - it applies both filters and transformations as well as makes assertions on parameters. It's more like a general wrapper that allows interception.

**Comment by danderson00** [comment]
Closing this. All the data hook stuff is now tracked in the datahooks branch in the main repo.

**Pull request closed without merging**